### ISSUE TITLE
Assembly is not Strongly Named

### ISSUE THREAD

**Issue by gjorchard** [wis3guy/HalClient.Net#3]
I have started using HalClient and like it so far. However it is not Strongly Named which is a pain since my assembly that references it is.

**Comment by wis3guy** [comment]
@gjorchard I understand and agree that that would make the library better. I will update this shortly (within the coming week).

**Comment by wis3guy** [comment]
@gjorchard I have looked into this and even though it is relatively easy to sign the assembly, the problems arise when packaging up the nuget package. HalClient.NET actually makes use of Tavis.UriTemplates (https://github.com/tavis-software/Tavis.UriTemplates) which is also not strongly named.

This actually got me thinking about whether i really want the assembly to be signed or not. Reading this post, left me unsure: http://blog.marcgravell.com/2014/06/snk-we-need-to-talk.html

I suggest you clone the repository and sign the assembly yourself for the time being.

**Comment by gjorchard** [comment]
Thanks for letting me know. I can certainly see where the author of the blog is coming from. 

It wasn't clear to me what the downside would be for you in making HalClient Strongly Named, but Tavis.UriTemplates would need to do the same to really help me. These are the only two packages, of the many being used here, that are not Strongly Named so it seems to be quite common.
I am working with code that only recently took over so I do not know what the decision was made to go with strong names. I will try to figure this out.
For now I have added strong names to both packages so I can continue to use them.
Regards,Greg

**Comment by darrelmiller** [comment]
@gjorchard Microsoft have recently updated their guidance on strong naming.  See [here](https://msdn.microsoft.com/en-us/library/wd40t7ad%28v=vs.110%29.aspx).  The main reason was that people mistakenly believed that strong naming brought certain security benefits.  That is not the case.  However, there are some identity related benefits if the library is used in plug-in environments.  The challenge for OSS providers is what to do with the private key used to generate the identity.  Microsoft are currently recommending that OSS developers publish the private key along with their source code in public repositories.  As you can see, security is definitely not a benefit.

There is much debate in the OSS community as to whether to accept requests to SN, or to reject them.  If the requests are motivated by a false sense of security, then possibly it is better to reject them.  However, if the end-user is trying to build a plug-in for Visual Studio then the request is possibly justified.

There has been much past pain caused by Strong Names relating to dependency management.  I believe the current guidance avoids that pain, but there is still reluctance among many to allow Strong Names to propagate through OSS ecosystem without significant justification.

**Comment by wis3guy** [comment]
@darrelmiller Thank you for this elaborate response. Based on what i read on Marc Gravell's blog, combined with the resources you provided, i feel it just does not make sense to sign the HallClient.NET assembly. If people feel they do need a signed version, they can clone the repository and sign it themselves.

**Comment by gjorchard** [comment]
Hi Geoff,
I am still using the 1.2 version of the library and have been very happy with it. Recently an issue has arisen that I want to bounce off of you. The API I am integrating with, in response to a POST to add an item to a list, returns a 201 (CREATED) with no body. Since there is no body they do not supply a content-type in the header. The HalClient balks a this throwing an error. While they do not supply a body they return a link to the newly created object in the location header. From a quick look I don't think you are providing access to the header via HalClient at the moment. Is this correct or am I missing something? Of course it is harder to look for when there is a exception thrown every time I receive a message with the location header set. :)
Is there anything in the current version of the library that would help me with these issues? If not what are your thoughts on them?
Regards,Greg

**Comment by wis3guy** [comment]
This is actually quite common for API's to do. In such cases, you could use the `HttpClient` property on the `IHalHttpClient` instance, like so:

```c#
public static async Task<Uri> GuardedPostAsJsonAsync<T>(this HttpClient client, Uri uri, T model)
{
	if (client == null) throw new ArgumentNullException(nameof(client));
	if (uri == null) throw new ArgumentNullException(nameof(uri));
	if (model == null) throw new ArgumentNullException(nameof(model));

	using (var response = await client.PostAsJsonAsync(uri, model))
	{
		if (!response.IsSuccessStatusCode)
			throw await response.AsApiRequestExceptionAsync();

		return response.Headers.Location;
	}
}
``` 

NB. `PostAsJsonAsync` is an extension method, provided by Microsoft.

**Comment by gjorchard** [comment]
Thanks Geoff.
I have tried this and am stuck on one thing. The HttpClient property on the IHalHttpClient is of type INonParsingHttpClient, not HttpClient the below is not working for me. 

What am I missing?
Greg

**Comment by wis3guy** [comment]
Hi Greg,

That is because you are using an older version. In the latest version the
wrapped HttpClient is exposed directly.

Is there a reason you haven't upgraded yet?

Ciao,
Geoffrey

**Comment by gjorchard** [comment]
I was hoping that was the reason!

I am just starting a new version of my tool set so upgrading now makes sense. I just did so and everything continues to work with the new version. Using the HttpClient within the HalClient now works as you said.

Why does the HalClient raise a NotSupportedException if there is no content-type and no body? While using the HttpClient directly would avoid this issue, it is not a great solution for me as one of my tools does not know whether a body is expected or not in the response. It is a test tool where GETs are fully coded to, but I use reflection in calling most PUTs and POSTs and then only look at the return code of the response, not the body. This isn't causing me much grief as I simply catch and ignore these exceptions but I would rather not in case an exception of this type is raised for some other reason at some point.
I think there is just one case so far where I need to look at the location header so I will use the HttpClient there.

Thanks again,Greg

**Comment by wis3guy** [comment]
Hi Greg,

> Why does the HalClient raise a NotSupportedException if there is no content-type and no body?

That is because the client is nothing more than a specialised wrapper around the `HttpClient`. It seems unlikely that API's would return `application\hal+json` for some responses and something else/nothing for others without the client being able to find out about this. Tbh i think it would not be wise to cater for inconsistent/mal-documented API's in the client.

What you seem to want is a client that silently fails parsing, correct? What about those scenario's where a 'consistent API' returns an unexpected non-HAL response? In those cases you would want an exception imho.

My suggestion to you would be to flip this around. Default to the regular HTTP client and _only_ use the HalHttpClient in case you are sure you'll receive `application\hal+json`.

**Comment by gjorchard** [comment]
Geoffrey,

My understanding from talking the to the web service developer is that this is the behavior when using JAX-RS. If a response has no body, the JAX-RS framework does not include a header-type. Intuitively this makes sense to me. Why provide a type for something that isn't there?
I am off until the new year but will collect any details I can find and provide them to you when I am back.
Regards,Greg

**Comment by wis3guy** [comment]
Hi Greg,

This code comes from the `HallHttpClient` class:

```c#
private async Task<IRootResourceObject> ProcessResponseMessage(HttpResponseMessage response)
{
	if ((response.StatusCode == HttpStatusCode.Redirect) ||
		(response.StatusCode == HttpStatusCode.SeeOther) ||
		(response.StatusCode == HttpStatusCode.RedirectMethod))
		return await GetAsync(response.Headers.Location);

	string mediatype = null;
	var isHalResponse = false;

	if (response.Content.Headers.ContentType != null)
	{
		mediatype = response.Content.Headers.ContentType.MediaType;
		isHalResponse = mediatype.Equals(ApplicationHalJson, StringComparison.OrdinalIgnoreCase);
	}

	if (response.IsSuccessStatusCode)
	{
		if (response.StatusCode == HttpStatusCode.NoContent) // SEE, THIS IS ALREADY HANDLED BY THE CLIENT ...
			return new RootResourceObject();

		if (string.IsNullOrEmpty(mediatype))
			throw new NotSupportedException("The response is missing the 'Content-Type' header");

		if (!isHalResponse)
			throw new NotSupportedException("The response contains an unsupported 'Content-Type' header value: " + mediatype);

		return await ParseContentAsync(response);
	}

	if (!isHalResponse)
		throw new HalHttpRequestException(response.StatusCode, response.ReasonPhrase);

	var resource = await ParseContentAsync(response);

	throw new HalHttpRequestException(response.StatusCode, response.ReasonPhrase, resource);
}
```

You stated that:

1. You received a `NotSupportedException`
2. The API returned a response without a `Content-Type` header
3. The API returned a response without a body 

My preliminary conclusion is that the API is returning an inconsistent status code. AFAIK the API should tell the client that there is no content by means of the HTTP 204 status code (or possibly a 304, which is not supported yet as you can see above). For any client to safely assume the response needs not be parsed, and that the empty response is not an error/problem, the status code should be set accordingly.

Could you tell me what status code is returned by the consumed API? Do you have any influence on the API you are consuming?

**Comment by gjorchard** [comment]
Hi Geoffrey,
Thanks for this. I will discuss the possible use of the 204 return code with the developer of the API. So yes, I do have influence over the API.

I believe the only status codes I am seeing this behavior with are 200 and 201.

I'll get back to you in the new year once I have discussed this with the developer.
Greg

**Comment by gjorchard** [comment]
Hi Geoffrey,
Below is the response from the designer of the web service. 

Let me know your thoughts.     

      
Message body
  
 Here is what I prefer from: https://specs.openstack.org/openstack/api-wg/guidelines/http.html    For POST it should be 201 with a location header. Should the response include the new created content? It is “SHOULD” not “MUST” in all HTTP standard(HTTP POST or status code 201). So from our point of view, I prefer to include some content in the body, BUT, the JAX-RS standard(java REST standard), when you do Response.created, its implementation does not include a body. So from a library point of view, it should support the scenario that client did not provide content body, because it is “SHOULD”.    204 should be used in DELETE, I think. 2xx Success Codes¶    
   - Synchronous resource creation
 ·        Response status code must be201Created ·        Must return a Location header with the URI of the created resource ·        Should return a representation of the resource in the body    
   - Asynchronous resource creation       
      - Response status code must be202Accepted
      - Must return a Location header set to one of the following:          
         
         - the URI of the resource to be created, if known.
         - the URI of a status resource that the client can use to query the progress of the asynchronous operation.


   
   - Synchronous resource deletion
 ·        Response status code must be204NoContent    
   - For all other successful requests, the return code should be200 OK.
   - If a request attempts to put a resource into a state which it is already in (for example, locking an instance which is already locked), the return code should be in the2xx Successful range (usually matching the return code which would be given if the state had changed). It is not appropriate to use409 Conflict when the resulting state of the resource is as the user requested.
       From: Orchard, Greg
Sent: Tuesday, December 20, 2016 9:09 AM
To: Xing, Steven
Subject: Handling of responses with no content    Steven,    Below is an email from Geoffrey Braaf, the author of the HalClient package we are using. It is his assertion that the correct return code to use on a successful response when there is no body is code 204 (NoContent) rather than 200 or 201 as we use today. As he indicates below, HalClient handles 204 correctly.    https://httpstatuses.com/204    What do you think? 
----- Forwarded Message -----

**Comment by wis3guy** [comment]
Fair point regarding the MUST vs SHOULD in the HTTP spec.

Let me think about it before i decide how to move forward. An additional concerns is also that in the current version you do not have access to the returned Location header. Adding this to the response object elegantly is a breaking change.

**Comment by gjorchard** [comment]
Thanks Geoffrey, let me know what you decide.
For me the handling of no content in a non-204 response is more important than having access to the location header at this point.
Regards,Greg

**Comment by wis3guy** [comment]
@gjorchard Have a look at: https://github.com/wis3guy/HalClient.Net/tree/no-content-no-parsing

Note that this release will be a breaking change, as I now return a `HalHttpResponseMessage`!

**Comment by gjorchard** [comment]
I'll try to take a look at it when I am back at work on Jan 3.
Thanks a lot for doing this!!
Greg

**Comment by gjorchard** [comment]
Hi Geoffrey,
I have partially integrated and all is good. I am not using the Location header yet as that will require more work on my side to use, but I see that it is available when I need it.
Let me know when you are going to release a version of the API with this in it.
Regards,Greg

**Comment by gjorchard** [comment]
Ended up having time to integrate with the Uri being in the location header. Works great. Much better than that work around I had in place.
Regards,Greg

**Comment by wis3guy** [comment]
I'll push to nuget sometime next week.

Ciao,
Geoffrey

**Comment by wis3guy** [comment]
@gjorchard The nuget package is now published and has version 2.0.0.

We should have done this in a separate issue ...
